src/sql/workbench/contrib/notebook/browser/outputs/gridOutput.component.ts

```diff
@@ -146,6 +147,48 @@ export class GridOutputComponent extends AngularDisposable implements IMimeCompo
 	}
 }
 
+function reorderGridData(source: IDataResource): void {
+	// Get Column Names list from the data resource schema
+	const columnNames: string[] = source.schema.fields.map(field => field.name);
+	if (source.data.length > 0) {
+		// Check to see if data source is ordered properly based on schema
+		// Papermill executed notebooks with KQL for instance will organize the
+		// row data in alphabetical order as a result the outputted grid will be
+		// unordered and not based on the data resource schema
+		let rowKeys = Object.keys(source.data[0]);
+		if (!equals(columnNames, rowKeys)) {
+			// SQL notebooks does not use columnName as key (instead uses indices)
+			// Indicies indicate the row is ordered properly
+			// We check the data to know if it is in index form or string form
+			for (let rowKey in rowKeys) {
+				// Check to see if first row keys are in index format (numbers)
+				let currentIndex = Number(rowKeys[rowKey]);
+				let nextIndex = Number(rowKeys[rowKey]) + 1;
+				// We are only testing to make sure that the entire row is in index format
+				// We assume that index format data keys are in order based on the data resource schema
+				// Edge case: If the index format is out of order, then we would not have the ability to match
+				// random indexes with the column names list
+				if (typeof currentIndex === 'number' && !Number.isNaN(currentIndex)) {
```

Comment:
What happens if we have 5 rows that have been misordered and the first two in the new "order" are numbers but the rest are strings? I think we should just be doing a single check at the beginning to see if we need to re-order or not (by checking if all of the keys are numbers and skipping if they are).

Reply:
And in fact right now if I'm reading this correctly you're re-ordering the entire table for every column which doesn't seem great perf wise. So I think it should look like this :

```
if(!equals(rowKeys, columnNames) {
    // check if all row keys are numbers, if they are then return immediately
    source.data.forEach((row, index => { 
         ...
    });
}
```

Reply:
This is what the statement does though (if condition), checks to see if all the keys are numbers. Are you saying to just put the `source.data.forEach((row, index => { `
in above the condition that checks for numbers?

Reply:
No, the opposite. Check whether the entire array is numbers first and if it is no-op. What you have right now is at best inefficient and at worst wrong. Two scenarios I see problems in : 

1. rowKeys = ["col1", "col2", "col3"]

What happens : 

For every key in that list (column) it goes through and does the source.data.forEach loop. This reorders the ENTIRE table of data every time which is unnecessary.

2. rowKeys = ["1", "col2", "col3"]

What happens :

You go into the loop and check the first two values. It sees the first is a number and the second isn't ("mixed" values) and so returns without reordering. But this doesn't seem right - if it's mixed we should be assuming that the values are all the actual column names and not indices and so do the reordering.

Reply:
Refactored the code and I believe addresses the efficiency of condition and reordering 👍